### DISCUSSION TITLE
Tips to cross reference assembled texture file with original camera images

### DISCUSSION THREAD

**Discussion by oppy2** [introlab/rtabmap#1052] [How to]
Hello all! 

Love this project, if anyone is able to point me in the right direction I'd be very thankful.

I am looking to be able to export the assembled mesh using pretty much the standard settings for iOS- so I end up with a Poisson mesh and the texture is generated from each photo being projected onto the mesh. My understanding (which may be faulty!) is:

Some photos (=cameras in the code) will be filtered out during the process and at the end a mesh is created, lets assume with only 1 texture. this texture, though, is actually made up of the photos scaled and arranged in the generated texture image. 

What I want to be able to do is to cross-reference from the generated (assembled, single texture) image file to the original individual images that were captured (in full resolution). If anyone has any starter tips for this that would be awesome.

Ideally a point in the code where I would be able to say here is the assembled image, and these are the photos that were used, and where they map to (scaled and translated) in the final texture.

Many thanks!
Oppy

**Comment by matlabbe** [comment]
Hi Oppy,

The number of subimages in the final texture may be different than the number of input camera images, though you may confirm with the parameters you are using. If a camera doesn't have any textured polygon, it will be skipped. The `materialsKept` variable here will tell you if an image has been kept or not.
https://github.com/introlab/rtabmap/blob/d031d7936930f127b03a972ce849d0941805fed7/corelib/src/util3d_surface.cpp#L1640-L1641

With outputs of `util3d::concatenateTextureMaterials`, you can recover which subimage in the texture belongs to which camera image index/id.

cheers,
Mathieu

**Reply by oppy2** [reply]
Hi Mathieu, that is exactly what I am after, thank you so much for the quick and detailed response! Cheers, Oppy.
